### ISSUE TITLE
 Unimplemented type for cast (TIMESTAMP WITH TIME ZONE -> TIMESTAMP_NS)

### ISSUE THREAD

**Issue by tekumara** [duckdb/duckdb#7980]
### What happens?

Cannot insert a timestamp with time zone into a timestamp(9) column.

### To Reproduce

```
❯ duckdb
v0.8.1 6536a77232
Enter ".help" for usage hints.
Connected to a transient in-memory database.
Use ".open FILENAME" to reopen on a persistent database.
D CREATE TABLE EXAMPLE (UPDATE_AT_NTZ TIMESTAMP(9));
D insert into example values (current_timestamp);
Error: Conversion Error: Unimplemented type for cast (TIMESTAMP WITH TIME ZONE -> TIMESTAMP_NS)
```

### OS:

darwin

### DuckDB Version:

v0.8.1 6536a77232

### DuckDB Client:

cli

### Full Name:

Oliver Mannion

### Affiliation:

Xero

### Have you tried this on the latest `master` branch?

- [X] I agree

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] I agree

**Comment by hawkfish** [comment]
Thanks - this is probably just an oversight. The cast is not lossy and we should just do it.

**Comment by hawkfish** [comment]
Yeah this works for regular timestamps, so ICU just needs to add it for TSTZ:

```sql
D create table issue7980 (ts TIMESTAMP(9));
D insert into issue7980 values (current_timestamp::TIMESTAMP);
D
```

**Comment by hawkfish** [comment]
I had forgotten that per the standard, this will create a naïve timestamp:

```sql
D create table issue7980 (ts TIMESTAMP(9));
D insert into issue7980 values (current_timestamp::TIMESTAMP);
D from issue7980;
┌─────────────────────────┐
│           ts            │
│      timestamp_ns       │
├─────────────────────────┤
│ 2023-06-19 07:59:20.777 │
└─────────────────────────┘
```

So I can add this but it may not do what you expect (current time when that query was run was `2023-06-19 14:59:20.777+00`).

**Comment by hawkfish** [comment]
I would suggest that you insert the value `current_timestamp AT TIME ZONE 'UTC'` to avoid the naïve timestamp madness.

**Comment by hawkfish** [comment]
I'd prefer that we not "fix" this as I consider the standard's TSTZ => TS behaviour to be broken and dangerous.

**Comment by tekumara** [comment]
Thanks for looking into this.. We are using nanoseconds to adjust the order of CDC rows recorded at the same millisecond timestamp.

btw, if TIMESTAMP WITH TIME ZONE => TIMESTAMP is broken, does that mean that `select current_timestamp::TIMESTAMP` should fail too?
 
Just trying to reconcile the fact that TIMESTAMP WITH TIME ZONE => TIMESTAMP and TIMESTAMP => TIMESTAMP_NS are possible, but TIMESTAMP WITH TIME ZONE => TIMESTAMP_NS isn't.

**Comment by hawkfish** [comment]
By "broken" I mean "poorly thought out by the standards committee" ;-). 

The cast from `TIMESTAMP WITH TIME ZONE` => `TIMESTAMP` creates a naïve timestamp, which may not be what you - or anyone else  - really wants because it can have gaps and ambiguous values.  If you are looking for ns accuracy, that is probably not what you want, and simply implementing the cast would not solve the problem. My suggestion is that you `Set TimeZone='UTC'` and then do a double cast (`current_timestamp::TIMESTAMP::TIMESTAMP_NS`).

Is the problem that the incoming data is in `TSTZ` columns and you want to uniquify them by adding a ns offset? 

One thing that is missing is the inverse of the (apparently undocumented - sorry!) function `epoch_ns`, which will extract the count of ns since the epoch. `make_timestamp(µs)` does exist and maybe what is really needed here is `make_timestamp_ns`?
